### PR TITLE
[RSI, performance] Land the kernel skill-sync marker so killed sessions do not re-pay it

### PR DESCRIPTION
**Pull request by sethkarten**
## What

- The kernel Python bootstrap now writes its venv `.bootstrap-version` marker atomically (temp file + rename in the same dir, direct-write fallback) instead of one plain write at the very end of the skill sync.
- `bootstrapVenv` persists the base marker (schema, runtime identity, snapshot, extra uv args) immediately after the runtime install, before the Python-skill sync.
- `syncPythonSkills` persists the marker after every successfully installed skill group, so the marker always reflects completed work. The final unconditional write is kept, so install-failure semantics are unchanged: a failed skill warns, continues, and retries next startup; the warning is now exclusively for real install failures (marker-write failures propagate as bootstrap errors, like the final write on main).
- No behavior change to the ready gate (exact marker match still short-circuits with zero uv), the per-skill skip (pyproject path + hash match), `PRIME_AGENT_KERNEL_PYTHON` overrides, legacy-schema rebuilds, or the bootstrap dir-lock.

## Why

PR #2398's idle-RSS profile caught `uv pip install --editable` re-running the bundled Python-skill sync inside the measurement window in 10/10 trials (~140 MB uv resident at idle). Root cause: the marker was written only once, after the LAST skill install, and no base marker existed before the skill sync - so any session killed mid-sync (the benchmark, and real short-lived sessions) recorded nothing. The next startup then failed `kernelBaseReady` and took the full wipe path: `rm -rf` the venv, re-install Python + prime-agent-runtime + extras, and re-run every editable skill install. Short-lived sessions re-pay this forever; the sync never converges.

With this change, a killed session leaves a valid base marker plus the skills that already finished. The next session takes the skills-only path and installs only the remainder; once any session completes the sync, all later sessions hit the ready gate with zero uv invocations.

## Local before/after

Same protocol as the benchmark harness: real uv 0.12.12, the real bundled release skills (11 with pyproject.toml), a uv shim logging every invocation, a throwaway venv, kill the process group mid-sync after 3 editable installs, then rerun to completion in a fresh process. All polls with explicit deadlines.

| metric | main | branch |
| --- | --- | --- |
| venv rebuild in rerun after kill | YES (`rm -rf` + `venv --seed` + runtime install) | NO (0 seed lines) |
| editable installs in rerun | 11 (all re-paid) | 9 (2 recorded at kill; remainder only) |
| peak uv RSS during rerun | 67.1 MB (70.6 MB concurrent) | 52.8-53.5 MB |
| kill-rerun wall time (warm uv cache) | 3.25 s | 2.30-2.62 s |
| cold full sync (sanity) | 3.27 s / 11 installs | 3.06 s / 11 installs |
| steady state after a completed sync | 0 uv | 0 uv |

Cumulative short-session demo (one venv, 6 consecutive sessions killed mid-sync): on the branch the marker grows 2->4->6->8->10 skills across sessions, the venv is never rebuilt, the 6th session completes the sync, and later sessions run zero uv. On main every killed session wipes the venv and re-pays the whole bootstrap; the sync never converges.

The warm-cache wall-time delta is modest; the structural win (no venv wipe, no runtime+extras reinstall, only the missing skills) scales with cache-cold rebuild cost, which is the real-world regression the #2398 profile caught (~140 MB uv resident per session, 10/10 trials).

## Testing

- `kernel-bootstrap.test.ts`: 27/27 green (23 pre-existing + 4 new):
  - "writes the version marker before syncing Python skills" - the fake uv probes the marker at every editable install; verified failing on pristine main (every probe sees no marker because main writes only after the last install).
  - "resumes a sync interrupted before all skills completed" - the exact post-kill marker state resumes without a rebuild and skips the completed skill.
  - "resumes a real killed mid-sync process without rebuilding the venv" - a real tsx child is SIGKILLed (process group) mid-install; the rerun steals the stale bootstrap lock via PID liveness, skips the completed skill, and never rebuilds the venv.
  - "re-syncs a changed skill alongside unsynced skills" - a changed pyproject re-syncs while an unsynced skill installs, with no rebuild.
- Other suites importing the kernel bootstrap: ipython-provisioner (21), repl-kernel-parent-watchdog (8 + 2 skipped), kernel-windows-process (6), binary-runtime-source (4), kernel-bootstrap-windows (2) - all green.
- `npm run check` (biome + tsgo + installer + browser-smoke) passes.

---

> [!NOTE]
> **Medium Risk**
> Changes kernel bootstrap persistence and skill-sync failure handling on disk; incorrect marker logic could cause unnecessary venv rebuilds or stale skill manifests, though atomic writes and existing ready gates limit worst-case impact.
> 
> **Overview**
> Fixes short-lived or killed sessions **rebuilding the kernel venv and re-running the full Python skill sync** because `.bootstrap-version` was only written once at the end of sync (or could be truncated mid-write).
> 
> **`bootstrap.ts`** now lands a **base marker** (runtime identity, empty skill list) right after the runtime install and **before** editable skill sync. Each successful skill install triggers **`writeMergedBootstrapVersion`**, which merges progress into the on-disk record; the final pass still does an authoritative **`writeBootstrapVersion`**. Marker updates use **write-temp + rename** with bounded retries so a crash never leaves a corrupt partial file that looks “missing” and forces a wipe.
> 
> Failed editable installs are **not** recorded (only appended after success); the loop **continues** so later skills can still install. Tests add fake-uv marker probes, a fail-path assertion, and a **real SIGKILL mid-sync** child process that proves the next startup skips completed skills and never runs `uv venv` again.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f7b547aaf33f71c3441cbaa9d092c84d0289f0a5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Write skill-sync marker atomically and incrementally in `bootstrapVenv`
> - `writeBootstrapVersion` in [bootstrap.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2405/files#diff-e87b9a04ab9b4ac02e8dd33af2f2f58320d8b7f6be6853acf4a3e05a34623cae) now serializes to a temp file and renames into place, retrying 3 times with a 50 ms delay so the prior marker is preserved on total failure.
> - New `writeMergedBootstrapVersion` helper merges existing marker skills with newly installed ones, allowing progress to persist across sync interruptions.
> - `bootstrapVenv` writes a base marker (empty skill list) right after runtime install and before Python skill sync, so a partially-synced venv is recognized as warm.
> - `syncPythonSkills` records each successful skill install incrementally and skips failed installs from the marker; the final write still replaces the full list authoritatively.
> - Added a child-process test helper and an integration test that kills a hanging skill install, then verifies resume behavior from the persisted marker.
> - Behavioral Change: failed skill installs are no longer recorded in the marker at any point; previously they could appear transiently before the final authoritative write.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized f7b547a.</sup>
> 
>

## Test-line budget

This PR must pass the repo's test-line-budget gate (`node scripts/check-test-policy.mjs`, run in CI): net added test lines may not exceed meaningful added source lines. Per the fleet-wide audit directive ("probably needs to remove redundant and unnecessary tests throughout. Remove unnecessary tests and we should have enough.") and the owner's follow-up ("nah dont grandfather them in. we should fix them now"), the branch compresses test lines by consolidation and removal of probe-verified duplicate coverage, with a lost-coverage ledger below.

- Before: net test additions 334 exceeded source additions 35, plus 6 policy-category violations (2 explicit per-test timeouts, vi.waitFor + wall-clock-sleep polls).
- After: net test additions 32 vs source 35 (gate passes; measured with the repo's checker against the PR merge-base). The two remaining policy annotations are rationaled on the real-kill test only (real killed-process semantics; no in-process signal exists).

**Lost-coverage ledger (executed cuts, probe-verified):** this branch's ledger documents NO distinct vector loss — every cut is either a consolidation or duplicate scaffolding, and each surviving pin was verified by mutating the behavior it must catch:

1. "writes the version marker before syncing Python skills" (branch test) merged into "writes the base marker before the first install and keeps a failed skill out of the record": same first-install marker probe (base marker before sync); the second-install incremental timing witness now lives in the kill test's at-kill marker. Probe: removing the base-marker write fails the merged test.
2. "resumes a sync interrupted before all skills completed" (branch test) removed: the resume observable (only the missing skill installs, no venv rebuild) is pinned by the kill test's rerun phase (hanging skill 2x, stale skill 1x, recorded skill 0x, zero venv lines).
3. The two real-kill tests merged into one warm-seeded kill test: the cold-kill variant's venv-line-count vector was redundant with the warm variant's zero-venv-line assert; the gated-admission machinery (vi.waitFor, 30s timeout, createRuntime gate) was scaffolding around vectors the ungated child pins. The merged test now kills via one group SIGKILL + exit-event await in the finally.
4. "re-syncs a changed skill alongside unsynced skills" (branch test) removed: the stale+missing combination is pinned by the kill test (the child pass reinstalls the stale skill; the rerun installs the missing one).
5. Pre-existing "syncs a warm venv when a Python skill pyproject changes" deleted: all three vectors (stale hash -> reinstall, fresh hash in the marker, no venv rebuild) survive in the kill test. Probe: making the stale check always-match fails the kill test at the at-kill fresh-hash assert.
6. Pre-existing "continues when a Python skill editable install fails and retries it next startup" deleted: fail-continue and not-recording survive in the new cold fail-probe test (probes: a rethrowing catch makes it reject; recording the failed skill fails the final-marker assert). Its retry segment duplicated the marker-driven resume decision pinned by the kill test's rerun.
7. Three bespoke fake-python builders (one warm/rebuild row, two PRIME_AGENT_KERNEL_PYTHON rows) consolidated into a `deniedProbes` parameter on the shared `writeFakePython` helper; every row asserts the identical error message as before.
- Merge-wave update (main advanced through #2399/#2409/#2410; this branch merges origin/main, keeping #2409's no-skill marker-preservation tests alongside this branch's compressed coverage): the gate re-measured against the new merge-base (main tip) is unchanged — net test additions 32 vs source 35, pass. #2409's two new tests are kept verbatim; the pre-existing "syncs a warm venv..." and "continues when a Python skill editable install fails..." deletions from this branch's audit remain probe-verified as covered by the merged kill/fail-probe tests.
- Merge-wave round 2 (main advanced through #2392, #2430, #2387; this branch merges origin/main again): #2387's no-seed venv change (drop `--seed` from `uv venv`) is kept together with this branch's durable-marker/atomic-swap coverage. The gate re-measured against the new merge-base (main tip fc9f2ae24) is unchanged — net test additions 32 vs source 35, pass. The pre-existing "syncs a warm venv when a Python skill pyproject changes" deletion remains: its no-rebuild vector is pinned by this branch's kill test with a seed-agnostic assertion, and its stale-reinstall/fresh-hash vectors by the kill test's child pass and at-kill assert.

### CODE REVIEW

File: packages/coding-agent/src/core/kernel/bootstrap.ts [hunk 1]
```diff
@@ -838,16 +853,20 @@ async function syncPythonSkills(
...
 			);
+			continue;
 		}
+		installedPythonSkills.push(
+			skill,
+			...localDependencies.filter((dependency) => !installedPythonSkills.includes(dependency)),
+		);
+		// Persist progress after every completed install so a killed session
+		// resumes at the first missing skill instead of re-syncing from scratch.
+		await writeBootstrapVersion(venv, runtimeIdentity, installedPythonSkills);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/kernel/bootstrap.ts]: Fixed in 27b432a15. The incremental write in `syncPythonSkills` now goes through a new `writeMergedBootstrapVersion` helper: it reads the on-disk marker (missing/corrupt tolerated as an empty base), unions the recorded skills with the accumulated list keyed by `${importName}\0${packagePath}` (fresh entries win, so a changed skill keeps its slot with its new hash), and writes through the existing `writeBootstrapVersion` (atomic tmp+rename preserved). Already-recorded skills later in install order now survive the incremental write, so a kill during a later install leaves them recorded and the next startup takes the skills-only path without re-paying unchanged editable installs. The final unconditional write remains an authoritative replace, so failed installs still retry and skills removed from the session list are still pruned.

Regression test: pre-seeded marker [A,B,C] with A's pyproject changed (only A installs as new), a brand-new skill D hangs its install and the child is SIGKILLed mid-D — the marker at kill is asserted [A,B,C] with A's new hash (pre-fix this left [A] only), and the resume performs 0 editable installs for B and C. Targeted kernel-bootstrap suite: 28/28 green; `npm run check` green.

File: packages/coding-agent/src/core/kernel/bootstrap.ts [hunk 2]
```diff
@@ -689,7 +690,34 @@ async function writeBootstrapVersion(
...
+	const serialized = `${JSON.stringify(version)}\n`;
+	try {
+		// Write-then-rename is atomic: a kill mid-write can never leave a
+		// partial marker, which would read as absent and force a rebuild.
+		const tmpPath = path.join(venv, BOOTSTRAP_VERSION_TMP_FILE);
+		await writeFile(tmpPath, serialized, "utf8");
+		await rename(tmpPath, filePath);
+	} catch {
+		await writeFile(filePath, serialized, "utf8");
+	}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/kernel/bootstrap.ts:717]: Fixed in fcb52b296: the truncating in-place fallback is gone, so the atomic-swap invariant now holds on every path. `writeBootstrapVersion` retries the swap (`BOOTSTRAP_MARKER_SWAP_ATTEMPTS = 3`, 50 ms apart, since replacing an existing marker can fail transiently while another process holds it) and, if it still fails, leaves the previous marker untouched and removes `.bootstrap-version.tmp`. Rationale in the code comment: a marker that stays stale only re-syncs skills on the next startup, while one truncated by a failed or killed overwrite reads as absent and forces a full venv rebuild.

Verified with a temporary test that mocks `node:fs/promises` `rename` to always reject, on the "resumes a sync interrupted before all skills completed" fixture: the marker file stays byte-identical, no `.bootstrap-version.tmp` is left behind, and the sync still installs the missing skill. The same test fails against the previous fallback, which rewrote the marker in place. Kept out of the commit because the branch is already over the test-line budget (net 328 test lines vs 27 source lines, inherited from the rebase wave).

File: packages/coding-agent/src/core/kernel/bootstrap.ts [hunk 3]
```diff
@@ -689,7 +694,46 @@ async function writeBootstrapVersion(
...
+			// Retry below; the previous marker is still intact.
+		}
+		if (attempt < BOOTSTRAP_MARKER_SWAP_ATTEMPTS) await sleep(BOOTSTRAP_MARKER_SWAP_RETRY_MS);
+	}
+	// Give up without touching the marker in place: a marker that stays stale
+	// only re-syncs skills, while one truncated by a failed or killed overwrite
+	// reads as absent and forces a full venv rebuild.
+	await rm(tmpPath, { force: true }).catch(() => undefined);
+}
+
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/kernel/bootstrap.ts:726]: Fixed in 378932c2 (commit amended). `writeBootstrapVersion` now records the last swap error and rethrows it after removing `.bootstrap-version.tmp`, so an unwritable marker surfaces through `ensureKernelPython`'s `formatBootstrapFailure` instead of leaving the venv silently unusable. The give-up path still never overwrites the marker in place, so the file cannot be left truncated and half-written.

Verified with a temporary test that mocks `node:fs/promises` `rename` to always reject on the "resumes a sync interrupted before all skills completed" fixture: `ensureKernelPython` rejects with `Failed to set up the Python kernel runtime. <EPERM ...>`, the marker file stays byte-identical, and no `.bootstrap-version.tmp` is left behind. Kept out of the commit because the branch is already over the test-line budget (net 328 test lines vs 27 source lines, inherited).

**Review by kevinjosethomas — APPROVED** [review]
